actors/miner/src/lib.rs

```diff
@@ -1950,6 +1958,306 @@ impl Actor {
+        let activation_epoch = rt.curr_epoch();
+
+        if params.proving_deadline >= policy.wpost_period_deadlines {
+            return Err(actor_error!(
+                illegal_state,
+                "proving deadline index {} invalid",
+                params.proving_deadline
+            ));
+        }
+
+        if !deadline_is_mutable(
+            policy,
+            state.current_proving_period_start(policy, curr_epoch),
+            params.proving_deadline,
+            curr_epoch,
+        ) {
+            return Err(actor_error!(
+                illegal_argument,
+                "proving deadline {} must not be the current or next deadline ",
+                params.proving_deadline
+            ));
+        }
+
+        if consensus_fault_active(&info, rt.curr_epoch()) {
+            return Err(actor_error!(
+                forbidden,
+                "prove-commit-ni not allowed during active consensus fault"
+            ));
+        }
+
+        rt.validate_immediate_caller_is(
+            info.control_addresses.iter().chain(&[info.worker, info.owner]),
+        )?;
+
+        let proof_inputs = params
+            .sectors
+            .iter()
+            .map(|sector| -> Result<SectorSealProofInput, ActorError> {
+                Ok(SectorSealProofInput {
+                    registered_proof: params.seal_proof_type,
+                    sector_number: sector.sealing_number,
+                    randomness: Randomness(
+                        rt.get_randomness_from_tickets(
+                            DomainSeparationTag::SealRandomness,
+                            sector.seal_rand_epoch,
+                            &entropy,
+                        )?
+                        .into(),
+                    ),
+                    interactive_randomness: Randomness(
+                        cid_to_replica_commitment_v1(&sector.sealed_cid)
+                            .map_err(|e| {
+                                actor_error!(
+                                    illegal_argument,
+                                    "invalid sealed CID {}: {}",
+                                    sector.sealed_cid,
+                                    e
+                                )
+                            })?
+                            .to_vec(),
+                    ),
+                    sealed_cid: sector.sealed_cid,
+                    unsealed_cid,
+                })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        verify_aggregate_seal(
+            rt,
+            // All the proof inputs, even for invalid pre-commits,
+            // must be provided as witnesses to the aggregate proof.
+            &proof_inputs,
+            params.sectors[0].sealer_id,
+            params.seal_proof_type,
+            params.aggregate_proof_type,
+            &params.aggregate_proof,
+        )?;
+
+        let base_sector_power = base_power_for_sector(info.sector_size);
```

Comment:
I wonder is it worth making a function here just for `BigInt::from`? Is this looking forward to some future iteration?

Reply:
I'm not aware of any future iterations, but I wanted to stay consistent with usage of `qa_power_for_sector`. Intent is more clear when reading the code.